internal/akams/akams.go

```diff
@@ -140,49 +172,101 @@ func (c *Client) CreateOrUpdateBulk(ctx context.Context, links []CreateLinkReque
 	switch resp.StatusCode {
-	case http.StatusAccepted, http.StatusNoContent, http.StatusNotFound:
-		// success
+	case http.StatusOK:
+		return true, nil
+	case http.StatusNotFound:
+		return false, nil
 	default:
-		return c.reqError(resp)
+		return false, c.reqError(resp)
 	}
-	return nil
 }
 
 func (c *Client) reqError(resp *http.Response) error {
+	// Try to preserve the response body. It may have important context to fix the issue.
 	body, _ := io.ReadAll(resp.Body)
 	return &ResponseError{StatusCode: resp.StatusCode, Body: string(body)}
 }
 
-func (c *Client) newRequest(ctx context.Context, method string, urlStr string, body any) (*http.Request, error) {
+func (c *Client) newRequest(ctx context.Context, method string, urlStr string, body io.Reader) (*http.Request, error) {
 	u := c.baseURL.JoinPath(urlStr)
-	var buf io.ReadWriter
-	if body != nil {
-		buf = &bytes.Buffer{}
-		enc := json.NewEncoder(buf)
-		enc.SetEscapeHTML(false)
-		if err := enc.Encode(body); err != nil {
-			return nil, err
-		}
-	}
-	req, err := http.NewRequestWithContext(ctx, method, u.String(), buf)
+	req, err := http.NewRequestWithContext(ctx, method, u.String(), body)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create request: %v", err)
 	}
 	req.Header.Set("Accept", "application/json")
 	if body != nil {
 		req.Header.Set("Content-Type", "application/json")
 	}
 	return req, nil
 }
+
+func (c *Client) do(req *http.Request) (*http.Response, error) {
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("request failed: %v", err)
+	}
+	return resp, nil
+}
+
+// chunkSlice chunks s into bulkSize chunks preserving the order of the items,
+// encodes each chunk into JSON, and calls fn with the encoded chunk.
+// The maximum encoded size of each chunk is limited to maxSizeBytes.
+// If the size of an item is larger than maxSizeBytes, an error will be returned.
+// If fn returns an error, the function will stop and return the error.
+func chunkSlice[T any](s []T, bulkSize int, maxSizeBytes int, fn func(io.Reader) error) error {
+	var buf bytes.Buffer
+	buf.WriteByte('[')
+	if len(s) == 0 {
+		buf.WriteByte(']')
+		return fn(&buf)
+	}
+	// We try to convert the slice to JSON in chunks to avoid hitting the maximum request size.
+	// The end result have the same encoding as if we had used json.Marshal.
+	enc := json.NewEncoder(&buf)
+	enc.SetEscapeHTML(false)
+	var callFn bool
+	for i := 0; i < len(s); {
+		lastSize := buf.Len() // keep in case we need to rewind.
+		if err := enc.Encode(s[i]); err != nil {
+			return err
+		}
+		buf.Truncate(buf.Len() - 1) // Remove the trailing newline added by enc.Encode.
+		buf.WriteByte(',')          // Add a comma to separate items.
+		if buf.Len() > maxSizeBytes {
+			// The last item was too big.
+			if size := buf.Len() - lastSize; size > maxSizeBytes {
+				// The last item is too big to fit in a chunk.
+				return fmt.Errorf("item %d is too large: %d bytes", i, size)
```

Comment:
Include max size to help the user see if they've accidentally configured this to something super small?
```suggestion
				return fmt.Errorf("item %d is too large: %d bytes > %d byte maximum", i, size, maxSizeBytes)
```